core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java

```diff
@@ -0,0 +1,247 @@
+                consumer.accept(spool.targetObject, v);
+            }
+            /*
+             * The target hasn't been built. Spool the consumer. The next two lines are the only allocations that ObjectSpooler does during
+             * parsing other than the boxing the ObjectParser might do. The first one is to preserve a snapshot of the current location so
+             * we can add it to the error message if parsing fails. The second one (the lambda) is the actual operation being spooled. Note
+             * that we don't do any of this if the object has already been built.
+             */
+            XContentLocation location = spool.parser.getTokenLocation();
+            spool.spool(targetObject -> {
+                try {
+                    consumer.accept(targetObject, v);
+                } catch (Exception e) {
+                    throw new ParsingException(location,
+                            "[" + objectParser.getName() + "] failed to parse spooled field [" + parseField.getPreferredName() + "]", e);
+                }
+            });
+        };
+    }
+
+
+    /**
+     * The target of the ObjectParser. One of these is built every time you call {@linkplain ObjectSpooler#apply(XContentParser, Object)}
+     * Note that it is not static so it inherits {@linkplain ObjectSpooler}'s type parameters.
+     */
+    public class Spool {
+        /**
+         * Array of constructor args to be passed to the {@link ObjectSpooler#builder}.
+         */
+        private final Object[] constructorArgs = new Object[constructorArgNames.size()];
+        private final XContentParser parser;
+        /**
+         * How many of the constructor parameters have we collected? We keep track of this so we don't have to count the
+         * {@link #constructorArgs} array looking for nulls when we receive another constructor parameter. When this is equal to the size of
+         * {@link #constructorArgs} we build the target object.
+         */
+        private int constructorArgsCollected = 0;
+        /**
+         * List of spooled attributes to be saved to the target object when we can build it. This is only allocated if an attribute has to
+         * be spooled.
+         */
+        private List<Consumer<Value>> spooled;
+        /**
+         * The target object. This will be instantiated with the constructor arguments are all parsed.
+         */
+        private Value targetObject;
+
+        public Spool(XContentParser parser) {
+            // TODO we probably don't need the parse here any more
+            this.parser = parser;
+        }
+
+        /**
+         * Set a constructor argument and build the target object if all constructor arguments have arrived.
+         */
+        private void constructorArg(int position, Object value) {
+            if (constructorArgs[position] != null) {
+                throw new IllegalArgumentException("Can't repeat constructor param");
```

Comment:
This message isn't going to mean a lot to the end user who reads it. I think we need to change it to `Can't repeat param [PARAM_NAME]` and pass the param name (`ParseField#getPreferredName()`) or `ParseField` itself into this method.